libgadget/lenstools.c

```diff
@@ -0,0 +1,336 @@
+}
+
+void grid3d_nfw(const struct particle_data * Parts, int num_particles, double **binning, GridDimensions dims, double *density) {
+    
+    double position[3];
+    // Process each particle
+    for (int p = 0; p < num_particles; p++) {
+
+        // remove offset
+        for(int d = 0; d < 3; d ++) {
+            position[d] = Parts[p].Pos[d] - PartManager->CurrentParticleOffset[d];
+            while(position[d] > PartManager->BoxSize) position[d] -= PartManager->BoxSize;
+            while(position[d] <= 0) position[d] += PartManager->BoxSize;
+        }
+        int ix = find_bin(position[0], binning[0], dims.nx, PartManager->BoxSize);
+        int iy = find_bin(position[1], binning[1], dims.ny, PartManager->BoxSize);
+        int iz = find_bin(position[2], binning[2], dims.nz, PartManager->BoxSize);
+
+        // continue if the particle is outside the grid
+        if (ix == -1 || iy == -1 || iz == -1) {
+            continue;
+        }
+        // Increment the density in the appropriate bin
+        ACCESS_3D(density, ix, iy, iz, dims.ny, dims.nz)++;
+    }
+
+}
+
+void projectDensity(double *density, double *density_projected, GridDimensions dims, int normal) {
+    // z; x, y
+    // y; z, x
+    // x; y, z
+    int DimNorm = (normal == 0) ? dims.nx : (normal == 1) ? dims.ny : dims.nz;
+    int Dim0 = (normal == 2) ? dims.nx : (normal == 0) ? dims.ny : dims.nx;
+    int Dim1 = (normal == 2) ? dims.ny : (normal == 1) ? dims.nz : dims.nz;
+
+    for (int i = 0; i < Dim0; i++) {
+        for (int j = 0; j < Dim1; j++) {
+            ACCESS_2D(density_projected, i, j, Dim1) = 0;
+            for (int k = 0; k < DimNorm; k++) {
+                if (normal == 0) {
+                    ACCESS_2D(density_projected, i, j, Dim1) += ACCESS_3D(density, k, i, j, dims.ny, dims.nz);
+                } else if (normal == 1) {
+                    ACCESS_2D(density_projected, i, j, Dim1) += ACCESS_3D(density, i, k, j, dims.ny, dims.nz);
+                } else {
+                    ACCESS_2D(density_projected, i, j, Dim1) += ACCESS_3D(density, i, j, k, dims.ny, dims.nz);
+                }
+            }
+        }
+    }
+}
+
+void calculate_lensing_potential(double *density_projected, int plane_resolution, double bin_resolution_0, double bin_resolution_1, double chi,double smooth, double *lensing_potential) {
+    // Allocate the complex FFT output array
+    fftw_complex *density_ft = fftw_malloc(sizeof(fftw_complex) * plane_resolution * (plane_resolution / 2 + 1));
+    // Initialize density_ft to zero
+    for (int i = 0; i < plane_resolution * (plane_resolution / 2 + 1); i++) {
+        density_ft[i][0] = 0.0;  // Real part
+        density_ft[i][1] = 0.0;  // Imaginary part
+    }
+
+    double *l_squared = allocate_2d_array_as_1d(plane_resolution, plane_resolution / 2 + 1);
+
+    // Create FFTW plans
+    fftw_plan forward_plan = fftw_plan_dft_r2c_2d(plane_resolution, plane_resolution, density_projected, density_ft, FFTW_ESTIMATE);
```

Comment:
I think that this is openmp parallel automatically because we call pfft_plan_with_nthreads before we do this

Reply:
Yes, I think it runs in omp parallel automatically already. I have compared with explicitly using omp here, the times spent are similar.